### DISCUSSION TITLE
Problem with TSPlayer.All

### DISCUSSION THREAD

**Discussion by Razrimo** [Pryaxis/TShock#2106] [Q&A] [answered]
Any reason why this code isn't working? When you do the command yourself it works
(Yes I know I could just set up the spawnpoint, this was just a test)

        public override void Initialize()
        {
            Commands.ChatCommands.Add(new Command(Telepe, "telepe")); 
            ServerApi.Hooks.ServerConnect.Register(this, OnJoin);
        }
        private void OnJoin(ConnectEventArgs args)
        {
               Commands.HandleCommand(TSPlayer.All, "/telepe");
        }
        void Telepe(CommandArgs args)
        {
            args.Player.Teleport(456 * 16, 105 * 16, 1);
        }

**Comment by Quinci135** [comment, marked as answer]
`TSPlayer.All` is a static readonly instance only for the sake of public broadcasts like `TSPlayer.All.SendInfoMessage("This is a public message.");`. Other methods do not work normally with this tsplayer object. TSPlayer constructer has the player index by default -1, and sendmessage will broadcast a message publicly if the player index is -1.
https://github.com/Pryaxis/TShock/blob/general-devel/TShockAPI/TSPlayer.cs#L1430
The `TShock.All.Kick` in Utils.cs doesn't work either (not that it matters, you get disconnected and notified of the server shutdown anyways. Probably due to an api change of kick being moved from utils to tsplayer and mistakenly replacing utils.kickall with tsplayer.all.kick)
https://github.com/Pryaxis/TShock/blob/general-devel/TShockAPI/Utils.cs#L468
Something like
```csharp
TSPlayer player = TShock.Players[args.Who];
player.Teleport(posX, posY, teleportStyle);
```
would work

Using the hook `ServerApi.Hooks.ServerConnect` would be too soon as then the player has yet to be spawned into the world nor would tshock have set a tsplayer object at that index. They are only connected to the server at this time. I'd NetGreetPlayer instead of ServerConnect.

If you use the Command constructor with those arguments `(Method, string[] commandNames)` then it will have no permissions and literally anyone can use that command. Constructor with permissions: `(List<string> permissions, CommandDelegate cmd, params string[] names)`
https://github.com/Pryaxis/TShock/blob/general-devel/TShockAPI/Commands.cs#L123

**Reply by Quinci135** [reply]
![ongreetconnect](https://user-images.githubusercontent.com/20407284/88862619-9ce8a700-d1b5-11ea-9d6d-a1a9ffaef9db.png)
```csharp
        private void OnGreet(GreetPlayerEventArgs args){
            TSPlayer player = TShock.Players[args.Who];
            if (player == null){
                Console.WriteLine($"A tsplayer was not found at index {args.Who}");
                return;
            }
            player.Teleport(Main.spawnTileX, Main.spawnTileY);
            Console.WriteLine("Attempted OnGreet teleport");
        }

        private void OnServerConnect(ConnectEventArgs args){
            TSPlayer player = TShock.Players[args.Who];
            if (player == null){
                Console.WriteLine($"A tsplayer was not found at index {args.Who}");
                return;
            }
            player.Teleport(Main.spawnTileX, Main.spawnTileY);
            Console.WriteLine("Attempted OnConnect teleport");
        }
```
With OnServerConnect tsplayer is null as tshock creates a tsplayer with the ServerConnect hook too
https://github.com/Pryaxis/TShock/blob/general-devel/TShockAPI/TShock.cs#L1164

**Reply by Razrimo** [reply]
Thanks!

**Comment by Razrimo** [comment]
Another question, if I were to do something similar in OnUpdate hook, how would I do it?

**Comment by Quinci135** [comment]
You can look at how tshock and other people's plugins do things for most books/interactions with the server and client
With tshock's on update
https://github.com/Pryaxis/TShock/blob/general-devel/TShockAPI/TShock.cs#L915

If you want to see the data in eventArgs each hook has you can use intellisense or look at https://github.com/Pryaxis/TerrariaAPI-Server/tree/general-devel/TerrariaServerAPI/TerrariaApi.Server/EventArgs

**Reply by Razrimo** [reply]
So I made this code that would randomly teleport everyone every couple of seconds, but for some reason it only affects the first player who joined the world

        private void OnUpdate(EventArgs args)
        {
            if (HasStarted == true)
            {
                Timer++;
                foreach (TSPlayer player in TShock.Players)
                {
                    if (Timer >= SetTime)
                    {
                        Timer = 0;
                        player.Teleport(Main.rand.Next(SizeMinX, SizeMaxX) * 16, Main.rand.Next(SizeMinY, SizeMaxY) * 16, 1);
                    }
                }
            }
        }

**Reply by Quinci135** [reply]
Why are you resetting the timer within the foreach loop? It will only execute once as the time condition is immediately reset.

**Reply by Razrimo** [reply]
If I reset the timer outside the foreach loop wouldn't that constantly reset therefore the if could never be achieved?

**Reply by Razrimo** [reply]
Managed to make it work in another way, thanks for all the help!
